subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdint.h>
+#include <zephyr.h>
+#include <flash.h>
+#include <download_client.h>
+#include <dfu/mcuboot.h>
+#include <dfu/flash_img.h>
+#include <pm_config.h>
+#include <logging/log.h>
+#include <fota_download.h>
+
+LOG_MODULE_REGISTER(fota_download, CONFIG_FOTA_DOWNLOAD_LOG_LEVEL);
+
+static 		fota_download_callback_t callback;
+static struct	flash_img_context flash_img;
+static struct	download_client dfu;
+
+static int download_client_callback(const struct download_client_evt *event)
+{
+	int err;
+
+	__ASSERT(event != NULL, "invalid dfu object\n");
+
+	switch (event->id) {
+	case DOWNLOAD_CLIENT_EVT_FRAGMENT: {
+		size_t size;
+
+		err = download_client_file_size_get(&dfu, &size);
+		if (err != 0) {
```

Comment:
I guess that could be a future improvement if we need it. 
For now the download will be canceled, and the application will not get any complete event.
This can again be handled by the application by a timeout.

Reply:
Yea, I suppose you have a point since the user won't get any event/return value when these happens, so they can't do anything about it. The download will stop, and that's pretty much it.